rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,845 @@
+  CBMC level quantifiers only allow certain kinds of expressions. This
+  necessitates a rewrite of the `Fn` closure to a compliant expression.
+- Letting the user supply the **harnesses for checking contracts** is a source of
+  unsoundness, if corner cases are not adequately covered. Ideally Kani would
+  generate the check harness automatically, but this is difficult both because
+  heap data structures are potentially infinite, and also because it must observe
+  user-level type invariants.
+
+  A complete solution for this is not known to us but there are ongoing
+  investigations into harness generation mechanisms in CBMC.
+
+  Environments that are non-inductive could be created from the type as the
+  safe Rust type constraints describe a finite space.
+
+  For dealing with pointers one applicable mechanism could be *memory
+  predicates* to declaratively describe the state of the heap both before and
+  after the function call. 
+  
+  In CBMC's implementation memory predicates are part of the pre/postconditions.
+  This does not easily translate to Kani, since we handle pre/postconditions
+  manually and mainly in proc-macros. There are multiple ways to bridge this
+  gap, perhaps the easiest being to add memory predicates *separately* to Kani
+  instead of as part of pre/postconditions, so they can be handled by forwarding
+  them to CBMC. However this is also tricky, because memory predicates are used
+  to describe pointers and pointers only. Meaning that if they are encapsulated
+  in a structure (such as `Vec` or `RefCell`) there is no way of specifying the
+  target of the predicate without breaking encapsulation (similar to
+  `modifies`).
+  
+  A better solution would be for the data structure to declare its own
+  invariants at definition site which are automatically swapped in on every
+  contract that uses this type.
+- What about mutable trait inputs (wrt memory access patters), e.g. a `mut impl AccessMe`
+- **Trait contracts:** Our proposal could be extended easily to handle simple
+  trait contracts. The macros would generate new trait methods with default
+  implementations, similar to the functions it generates today. Using sealed
+  types we can prevent the user from overwriting the generated contract methods.
+  Contracts for the trait and contracts on it's `impl`s are combined by abstracting
+  the original method depending on context. The occurrence inside the contract
+  generated from the trait method is replaced by the `impl` contract. Any other
+  occurrence is replaced by the just altered trait method contract.
+- **Cross Session Verification Caching:** This proposal focuses on scalability
+  benefits within a single verification session, but those verification results
+  could be cached across sessions and speed up verification for large projects
+  using contacts in the future.
+- **Inductive Reasoning:** Describing recursive functions can require that the
```

Comment:
Can you given an example of that ?

Reply:
The contract is attached to a function that is recursive, and the contract pre and post conditions can use other recursive predicates, but the contract itself recursing seems like a weird way to put it.